### ISSUE TITLE
Is anything about the GCNconv layer stochastic? - Manual GradientTape applying different weights even with the same gradients

### ISSUE THREAD

**Issue by amjass12** [danielegrattarola/spektral#378]
Hi @danielegrattarola , 

I am running a custom training loop to loop through individual graphs (going back to an older issue i had) - it actually all works fine, however i noticed that when i run `GradientTape()` twice, I get the **same** gradients, however the optimizer I am using (Adam) applies this differently each time and I get different weights applied - mostly to the `GCNconv` layer weights (and they are not floating point errors). 

As this is happening in the `GCNconv` layer, I was wondering if there is anything stochastic about this layer which may cause the optimizer to apply different weights each time, even though the gradients are identical? I apologise if this is a tensorflow issues- it is just that these weights are always different in the `GCNConv` layers specifically. Code below, but please note:

Gradient calculation identical EVERY time, so the gradients are being calculated correctly
The loss is also identical EVERY time, the mean loss does not change. 

`model.load_weights('weights.ckpt') #Weights identical every time`

```
with tf.GradientTape() as tape:  # Record operations for automatic differentiation.
       loss = #loss calculation 
    grads = tape.gradient(policy_loss, actor_model.trainable_variables)
    optimizer.apply_gradients(zip(grads, model.trainable_variables))

```

again, wheenver i load the same weights to the model - the gradients are identical every time - the loss is identical every time - but if i now look at `model.weights` the weights in the GCNConv layers are not the same each time! is this normal?

thank you!

**Comment by danielegrattarola** [comment]
Hi, 

I don't think this is a Spektral issue, unless when you load the model you get different weights than the ones you saved. Everything after that, it's most likely TensorFlow. 

In this case, Adam is a stateful optimizer so you need to make sure that you're restoring the same state if you want the same exact step. 

Cheers
